src/ScottPlot5/ScottPlot5/Legends/Legend.cs

```diff
@@ -25,48 +28,74 @@ public void Render(RenderPack rp)
         if (!IsVisible)
             return;
 
-        IEnumerable<LegendItem> allItems = rp.Plot.PlottableList.SelectMany(x => x.LegendItems).Concat(ManualItems);
-
-        LegendItem[] items = GetAllLegendItems(allItems).Where(x => x.IsVisible).ToArray();
-        if (!items.Any())
-            return;
-
         // measure all items to determine dimensions of the legend
         using SKPaint paint = new();
-        Font.ApplyToPaint(paint);
+        SizedLegendItem[] sizedItems = GetSizedLegendItems(rp.Plot, paint);
 
-        SizedLegendItem[] sizedItems = GetSizedLegendItems(items, paint);
+        if (sizedItems?.Any() != true)
+            return;
 
         float maxWidth = sizedItems.Select(x => x.Size.WithChildren.Width).Max();
         float totalheight = sizedItems.Select(x => x.Size.WithChildren.Height).Sum();
 
         PixelSize legendSize = new(maxWidth + Padding.Left + Padding.Right, totalheight + Padding.Top + Padding.Bottom);
         PixelRect legendRect = legendSize.AlignedInside(rp.DataRect, Alignment, Margin);
         PixelRect legendShadowRect = legendRect.WithDelta(ShadowOffset, ShadowOffset, Alignment);
+        Pixel offset = new(legendRect.Left + Padding.Left, legendRect.Top + Padding.Top);
 
         // render the legend panel
-        Drawing.Fillectangle(rp.Canvas, legendShadowRect, ShadowFill.Color);
-        Drawing.Fillectangle(rp.Canvas, legendRect, BackgroundFill.Color);
-        Drawing.DrawRectangle(rp.Canvas, legendRect, OutlineStyle.Color, OutlineStyle.Width);
+        RenderLegend(sizedItems, rp.Canvas, paint, offset, legendRect, legendShadowRect);
+    }
+    public void AsSvg(ScottPlot.Plot plot, Stream svgStream, int maxWidth = 0, int maxHeight = 0)
+    {
+        if (svgStream is null)
+            throw new NullReferenceException($"invalid Stream");
 
-        // render all items inside the legend
-        float yOffset = legendRect.Top + Padding.Top;
-        for (int i = 0; i < sizedItems.Length; i++)
+        var canvas = RenderToObject(plot, svgStream: svgStream, maxWidth: maxWidth, maxHeight: maxHeight) as SKCanvas;
```

Comment:
I'm really not a fan of `RenderToObject` being capable of returning two very different objects. I think you were on the right track with `RenderLegend`, which can take any canvas (whether an `SKSvgCanvas` or not).

Currently `RenderToObject` gives you either an `SKSurface` or an `SKCanvas` because it can do two different things. The code which is actually shared is the sizing code (measuring how big the legend should be) and the call to `RenderLegend`. The rest is creating either a surface or an SVG canvas, which isn't actually shared between the two usecases. So if we stop trying to share it (instead making the callers deal with that mess) we get a simpler implementation which is easier to reason about and to test.

What I would do is create a function called something like `MeasureLegend`, which returns `legendRect`, `maxWidth`, `maxHeight`, etc. Then the consuming functions do this:

- `Render` calls `MeasureLegend` and then calls `RenderLegend` with `rp.Canvas`
- `AsSvg` calls `MeasureLegend` and uses that to create an `SKSvgCanvas`. It passes that to `RenderLegend`
- `GetImage` calls `MeasureLegend` and uses that to create an `SKSurface`. It passes that to `RenderLegend`

Reply:
All of this is assuming we need SVG support for the legend, which I hadn't thought was necessary. If we don't then you can probably move the surviving code from `RenderToObject` back into `GetImage` and there'd be no problem.

Although I would still like the measurement code to be factored out. Otherwise it tends to get copy-pasted around and it can introduce lots of subtle bugs if one isn't very disciplined. If it's just a function that everyone calls rather than copy-pastes then that discipline is unnecessary.

Reply:
Thanks for this feedback @bclehmann! I'm going to mark this as resolved so I can merge now, but there's a lot to be desired about performance and design of the Legend class. Luckily the public API is simple and raster and vector exporting is supported now, and we can loop back around and refine the internals as time allows. I'll share some details in a comment at the bottom of this PR